File: 2015/001/001.c
#include <stdio.h>
#include <stdlib.h>

int getDestinationFloor(char *instructions, int instructionsLen) {
  int destinationFloor = 0;
  for (int i = 0; i < instructionsLen; i++) {
    char instruction = instructions[i];
    if (instruction == '(')
      destinationFloor++;
    if (instruction == ')')
      destinationFloor--;
  }
  return destinationFloor;
}

int getStartOfBasementPosition(char *instructions, int instructionsLen) {
  int floor = 0;
  for (int i = 0; i < instructionsLen; i++) {
    if (instructions[i] == '(')
      floor++;
    if (instructions[i] == ')')
      floor--;
    if (floor == -1)
      return i + 1;
  }
  return floor + 1;
}

int main(void) {
  FILE *f = fopen("input.txt", "r");
  int allocated = sizeof(char) * 100;
  char *instructions = malloc(allocated);
  int characterCount = 0;

  for (int c = fgetc(f); c != EOF; c = fgetc(f)) {
    instructions[characterCount] = c;
    characterCount++;
    if (sizeof(char) * characterCount >= allocated) {
      allocated *= 2;
      instructions = realloc(instructions, allocated);
    }
  }

  instructions = realloc(instructions, sizeof(char) * characterCount);
  int destinationFloor = getDestinationFloor(instructions, characterCount);
  int startOfBasement =
      getStartOfBasementPosition(instructions, characterCount);

  fclose(f);

  printf("The destination floor is %d\n", destinationFloor);
  printf("The start of basement is at position: %d\n", startOfBasement);
}

File: 2015/001/001.js
const fs = require("node:fs/promises");

/**
 * Calculates the destination floor based on the input string
 * @param {string} instructions
 * @returns {number}
 */
function getDestinationFloor(instructions) {
  let destinationFloor = 0;
  for (const instruction of instructions) {
    if (instruction == "(") {
      destinationFloor++;
      continue;
    }
    destinationFloor--;
  }
  return destinationFloor;
}

/**
 * Get the position of the instruction that first enters the basement, it's 1 indexed.
 * @param {string} instructions
 * @returns {number}
 */
function getStartOfBasementPosition(instructions) {
  let destinationFloor = 0;
  for (let i = 0; i < instructions.length; i++) {
    const instruction = instructions[i];
    if (instruction == "(") destinationFloor++;
    if (instruction == ")") destinationFloor--;
    if (destinationFloor == -1) return i + 1;
  }
  return -1;
}

(async function () {
  const input = await fs.readFile("input.txt", { encoding: "utf8" });
  const destinationFloor = getDestinationFloor(input.trim());
  const startOfBasement = getStartOfBasementPosition(input.trim());
  console.log("The destination floor is:", destinationFloor);
  console.log("The start of basement is at position:", startOfBasement);
})();

module.exports = {
  getDestinationFloor,
  getStartOfBasementPosition,
};
